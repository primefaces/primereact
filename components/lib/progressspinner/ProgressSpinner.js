import * as React from 'react';
import { classNames, ObjectUtils } from '../utils/Utils';

export const ProgressSpinner = React.memo(React.forwardRef((props, ref) => {
    const otherProps = ObjectUtils.findDiffKeys(props, ProgressSpinner.defaultProps);
    const className = classNames('p-progress-spinner', props.className);

    return (
        <div id={props.id} style={props.style} className={className} role="alert" aria-busy {...otherProps}>
            <svg className="p-progress-spinner-svg" viewBox="25 25 50 50" style={{ animationDuration: props.animationDuration }}>
                <circle className="p-progress-spinner-circle" cx="50" cy="50" r="20" fill={props.fill} strokeWidth={props.strokeWidth} strokeMiterlimit="10" />
            </svg>
        </div>
    )
}));

ProgressSpinner.displayName = 'ProgressSpinner';
ProgressSpinner.defaultProps = {
    __TYPE: 'ProgressSpinner',
    id: null,
    style: null,
    className: null,
    strokeWidth: '2',
    fill: 'none',
    animationDuration: '2s'
}
