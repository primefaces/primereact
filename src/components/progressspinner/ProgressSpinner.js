import React, {Component} from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

export class ProgressSpinner extends Component {

    static defaultProps = {
        id: null,
        style: null,
        className: null,
        strokeWidth: "2",
        fill: "none",
        animationDuration: "2s"
    };

    static propTypes = {
        id: PropTypes.string,
        style: PropTypes.object,
        className: PropTypes.string,
        strokeWidth: PropTypes.string,
        fill: PropTypes.string,
        animationDuration: PropTypes.string
    };

    render() {
        let spinnerClass = classNames('p-progress-spinner',this.props.className);

        return <div id={this.props.id} style={this.props.style} className={spinnerClass} role="alert" aria-busy={true}>
            <svg className="p-progress-spinner-svg" viewBox="25 25 50 50" style={{animationDuration: this.props.animationDuration}}>
                <circle className="p-progress-spinner-circle" cx="50" cy="50" r="20" fill={this.props.fill}
                        strokeWidth={this.props.strokeWidth} strokeMiterlimit="10"/>
            </svg>
        </div>;
    }
}
