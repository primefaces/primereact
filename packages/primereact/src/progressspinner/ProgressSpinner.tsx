'use client';
import { Component, withComponent } from '@primereact/core/component';
import { useProgressSpinner } from '@primereact/headless/progressspinner';
import { styles } from '@primereact/styles/progressspinner';
import { mergeProps } from '@primeuix/utils';
import * as React from 'react';
import { ProgressSpinnerProvider } from './ProgressSpinner.context';
import { defaultProps } from './ProgressSpinner.props';

export const ProgressSpinner = withComponent({
    name: 'ProgressSpinner',
    defaultProps,
    styles,
    setup(instance) {
        const progressspinner = useProgressSpinner(instance.inProps);

        return progressspinner;
    },
    render(instance) {
        const { id, props, ptmi, ptm, cx } = instance;

        const createSVGElement = () => {
            const circleProps = mergeProps(
                {
                    className: cx('circle'),
                    cx: '50',
                    cy: '50',
                    r: '20',
                    fill: props.fill,
                    strokeWidth: props.strokeWidth,
                    strokeMiterlimit: 10
                },
                ptm('circle')
            );

            const svgProps = mergeProps(
                {
                    className: cx('spin'),
                    viewBox: '25 25 50 50',
                    style: {
                        animationDuration: props.animationDuration
                    }
                },
                ptm('spin')
            );

            return (
                <svg {...svgProps}>
                    <circle {...circleProps} />
                </svg>
            );
        };

        const svg = createSVGElement();

        const rootProps = mergeProps(
            {
                id,
                className: cx('root'),
                role: 'progressbar'
            },
            ptmi('root')
        );

        return (
            <ProgressSpinnerProvider value={instance}>
                <Component instance={instance} attrs={rootProps}>
                    {svg}
                </Component>
            </ProgressSpinnerProvider>
        );
    }
});
