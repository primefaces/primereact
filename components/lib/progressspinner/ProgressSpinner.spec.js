import '@testing-library/jest-dom';
import { ProgressSpinner } from './ProgressSpinner';

import { snapshot } from '../../test';

describe('ProgressSpinner', () => {
    snapshot(<ProgressSpinner />, 'default');
    snapshot(<ProgressSpinner strokeWidth="3" />, 'strokeWidth');
    snapshot(<ProgressSpinner fill="green" />, 'fill');
    snapshot(<ProgressSpinner animationDuration="5s" />, 'animationDuration');
});
