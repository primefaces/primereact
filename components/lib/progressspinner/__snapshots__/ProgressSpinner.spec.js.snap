// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProgressSpinner animationDuration 1`] = `
<div>
  <div
    aria-busy="true"
    class="p-progress-spinner"
    role="alert"
  >
    <svg
      class="p-progress-spinner-svg"
      style="animation-duration: 5s;"
      viewBox="25 25 50 50"
    >
      <circle
        class="p-progress-spinner-circle"
        cx="50"
        cy="50"
        fill="none"
        r="20"
        stroke-miterlimit="10"
        stroke-width="2"
      />
    </svg>
  </div>
</div>
`;

exports[`ProgressSpinner default 1`] = `
<div>
  <div
    aria-busy="true"
    class="p-progress-spinner"
    role="alert"
  >
    <svg
      class="p-progress-spinner-svg"
      style="animation-duration: 2s;"
      viewBox="25 25 50 50"
    >
      <circle
        class="p-progress-spinner-circle"
        cx="50"
        cy="50"
        fill="none"
        r="20"
        stroke-miterlimit="10"
        stroke-width="2"
      />
    </svg>
  </div>
</div>
`;

exports[`ProgressSpinner fill 1`] = `
<div>
  <div
    aria-busy="true"
    class="p-progress-spinner"
    role="alert"
  >
    <svg
      class="p-progress-spinner-svg"
      style="animation-duration: 2s;"
      viewBox="25 25 50 50"
    >
      <circle
        class="p-progress-spinner-circle"
        cx="50"
        cy="50"
        fill="green"
        r="20"
        stroke-miterlimit="10"
        stroke-width="2"
      />
    </svg>
  </div>
</div>
`;

exports[`ProgressSpinner strokeWidth 1`] = `
<div>
  <div
    aria-busy="true"
    class="p-progress-spinner"
    role="alert"
  >
    <svg
      class="p-progress-spinner-svg"
      style="animation-duration: 2s;"
      viewBox="25 25 50 50"
    >
      <circle
        class="p-progress-spinner-circle"
        cx="50"
        cy="50"
        fill="none"
        r="20"
        stroke-miterlimit="10"
        stroke-width="3"
      />
    </svg>
  </div>
</div>
`;
